### PR TITLE
[FLINK-31233][Rest] Return 404 when task manager stdout/log files not exist.

### PR DESCRIPTION
**Pull request by huwh**
## What is the purpose of the change

When running Flink on Kubernetes, the stdout logs are not redirected to files so it will not shown in WEB UI. This is as expected.

But It returns “500 Internal error” in REST API and produces an error log in jobmanager.log. This is confusing and misleading.

## Brief change log
  - *Migrate some test to junit 5*
    - move TestingChannelHandlerContext out of AbstractTaskManagerFileHandlerTest
    - move TestingTaskManagerFileHandler out of AbstractTaskManagerFileHandlerTest
    - migrate AbstractTaskManagerFileHandlerTest and TaskManagerLogListHandlerTest to junit5
  - *remove the usage of mockito in TestingChannelHandlerContext*
    - create TestingChannelPipeline 
  - *return 200 and inform user to use kubectl logs to get stdout content*

## Verifying this change
  - *Add unit test to verify TaskManagerStdoutFileHandler handle FileNotFoundException and other exceptions.*
  - *Manually verified the change by running on kubernetes.*
  - give information when the stdout file does not exist.
image

  - Throw an exception when the log file does not exist. (manually deleted)
image

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (no)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (no)
  - The serializers: (no)
  - The runtime per-record code paths (performance sensitive): (no)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn, ZooKeeper: (no)
  - The S3 file system connector: (no)

## Documentation

  - Does this pull request introduce a new feature? (no)

### CODE REVIEW

**Comment by huwh** [comment]
@KarmaGYZ, could you please help review this PR in your free time? thanks a lot.

File: flink-runtime/src/test/java/org/apache/flink/runtime/rest/handler/taskmanager/TestingChannelHandlerContext.java [hunk 1]
```diff
@@ -0,0 +1,277 @@
...
+import java.io.File;
+import java.io.FileOutputStream;
+import java.io.IOException;
+import java.net.SocketAddress;
+
+/** Testing implementation of {@link ChannelHandlerContext}. */
+class TestingChannelHandlerContext implements ChannelHandlerContext {
+
+    final File outputFile;
+    final ChannelPipeline pipeline = new TestingChannelPipeline();
```
Comment (KarmaGYZ) [inline flink-runtime/src/test/java/org/apache/flink/runtime/rest/handler/taskmanager/TestingChannelHandlerContext.java:50]: Why not add private modifier here?

File: flink-runtime/src/main/java/org/apache/flink/runtime/rest/handler/taskmanager/TaskManagerStdoutFileHandler.java [hunk 2]
```diff
@@ -24,22 +24,33 @@
...
 
+import java.io.FileNotFoundException;
+import java.util.Collections;
 import java.util.Map;
 import java.util.concurrent.CompletableFuture;
 
 /** Rest handler which serves the stdout file of the {@link TaskExecutor}. */
 public class TaskManagerStdoutFileHandler
         extends AbstractTaskManagerFileHandler<TaskManagerMessageParameters> {
+    static final String FILE_NOT_FOUND_INFO =
```
Comment (KarmaGYZ) [inline flink-runtime/src/main/java/org/apache/flink/runtime/rest/handler/taskmanager/TaskManagerStdoutFileHandler.java:51]: Add `@VisibleForTesting` here.

File: flink-runtime/src/test/java/org/apache/flink/runtime/rest/handler/taskmanager/TestingChannelHandlerContext.java [hunk 3]
```diff
@@ -46,11 +48,21 @@ class TestingChannelHandlerContext implements ChannelHandlerContext {
 
     final File outputFile;
     final ChannelPipeline pipeline = new TestingChannelPipeline();
+    HttpResponse httpResponse;
+    byte[] responseData;
```
Comment (KarmaGYZ) [inline flink-runtime/src/test/java/org/apache/flink/runtime/rest/handler/taskmanager/TestingChannelHandlerContext.java:52]: Ditto. private modifier.

File: flink-runtime/src/test/java/org/apache/flink/runtime/rest/handler/taskmanager/TestingTaskManagerFileHandler.java [hunk 4]
```diff
@@ -49,6 +48,8 @@
 
     private final ResourceID expectedTaskManagerId;
 
+    final Exception exceptionForRequestFileUpload;
```
Comment (KarmaGYZ) [inline flink-runtime/src/test/java/org/apache/flink/runtime/rest/handler/taskmanager/TestingTaskManagerFileHandler.java:51]: private modifier.

File: flink-runtime/src/test/java/org/apache/flink/runtime/rest/handler/taskmanager/AbstractTaskManagerFileHandlerTest.java [hunk 5]
```diff
@@ -165,6 +168,81 @@ void testFileCacheExpiration() throws Exception {
...
+    }
+
+    @Test
+    void testStdoutFileHandlerHandleOtherException() throws Exception {
+        final Time cacheEntryDuration = Time.milliseconds(1000L);
+        TestingTaskManagerStdoutFileHandler testingTaskManagerStdoutFileHandler =
+                createTestTaskManagerStdoutFileHandler(
+                        cacheEntryDuration, new FlinkException("excepted exception"));
+        testHandleExceptionThrows(testingTaskManagerStdoutFileHandler);
+    }
```
Comment (KarmaGYZ) [inline flink-runtime/src/test/java/org/apache/flink/runtime/rest/handler/taskmanager/AbstractTaskManagerFileHandlerTest.java:206]: We might file a new `TaskManagerStdoutFileHandlerTest` for it.

File: flink-runtime/src/test/java/org/apache/flink/runtime/rest/handler/taskmanager/AbstractTaskManagerFileHandlerTest.java [hunk 6]
```diff
@@ -165,6 +168,81 @@ void testFileCacheExpiration() throws Exception {
...
+
+    @Test
+    void testDefaultHandlerHandleOtherException() throws Exception {
+        final Time cacheEntryDuration = Time.milliseconds(1000L);
+        final Queue<CompletableFuture<TransientBlobKey>> requestFileUploads = new ArrayDeque<>(1);
+        TestingTaskManagerFileHandler testTaskManagerFileHandler =
+                createTestTaskManagerFileHandler(
+                        cacheEntryDuration, requestFileUploads, EXPECTED_TASK_MANAGER_ID);
+        testHandleExceptionThrows(testTaskManagerFileHandler);
+    }
```
Comment (KarmaGYZ) [inline flink-runtime/src/test/java/org/apache/flink/runtime/rest/handler/taskmanager/AbstractTaskManagerFileHandlerTest.java:229]: Not sure whether these  two tests are necessary. We don't change the behavior of `AbstractTaskManagerFileHandler`.

**Review by KarmaGYZ — CHANGES_REQUESTED** [review]

**Comment by huwh** [comment]
Thanks for your review, @KarmaGYZ . Comments addressed, please take a look. 
The CI failed due to other issues, so I rebase it to the latest master branch. 

In AbstractTaskManagerFileHandler.java, i change the 'whenComplete' to 'handle' because 'whenComplete' could not swallow the exception. But it's strange to return an ignored variable. So i change it to return a 'null'.

**Comment by huwh** [comment]
Thanks, unit tests have been fixed.
`testStdoutFileHandlerHandleFileNotFoundException` failed because `TestingChannelHandlerContext#write` return an unfinished future. 
Since we already finish the write in test, we can return a finished future in `TestingChannelHandlerContext#write`

**Comment by huwh** [comment]
@flinkbot run azure

**Pull request merged**